### PR TITLE
Improved subdomain handling

### PR DESCRIPTION
**Pull request by fenek**
This PR partially addresses #911

Proposed changes include:
- `gen_mod` maintains a table of subdomain->domain mappings for modules that register themselves under these subdomains.
- Entries are automatically added/removed
- API is provided to manipulate module options by subdomain
- When any module uses `host` option, it is mandatory now (makes it easier to update mappings table)
- Main module table is now created with `read_concurrency` enabled
- Clarified naming convention in some places (Host/ServerHost/Server vs. Host/SubHost/MUCHost/etc.)
- Updated docs about `host` becoming mandatory
- `mod_mam_muc` is now configured with MUC module name, not domain; it also uses new registry instead of custom, private table
- `mod_muc_light` uses new registry instead of own, private table

### CODE REVIEW

File: apps/ejabberd/src/gen_mod.erl [hunk 1]
```diff
@@ -55,6 +57,12 @@
 -type ejabberd_module() :: #ejabberd_module{module_host :: {module(), ejabberd:server()},
                                             opts :: list()}.
 
+-record(ejabberd_module_subhost, {subhost, module, host}).
+-type ejabberd_module_subhost() :: #ejabberd_module_subhost{
```
Comment (michalwski) [inline apps/ejabberd/src/gen_mod.erl]: Would it be enough to name the type just `subhost`?

File: apps/ejabberd/src/mod_mam_muc.erl [hunk 2]
```diff
@@ -260,7 +225,7 @@ filter_room_packet(Packet, EventData) ->
         FromJID :: ejabberd:jid(), RoomJID :: ejabberd:jid(),
         Role :: mod_muc:role(), Affiliation :: mod_muc:affiliation()) -> packet().
 archive_room_packet(Packet, FromNick, FromJID=#jid{}, RoomJID=#jid{}, Role, Affiliation) ->
-    Host = server_host(RoomJID),
+    {_, Host} = gen_mod:get_module_and_host_by_subhost(RoomJID#jid.lserver),
```
Comment (michalwski) [inline apps/ejabberd/src/mod_mam_muc.erl]: I see there is many cases where only the `Host` is needed. Maybe it would be worth to create a function which returns just `Host`? `gen_mod:get_host_by_subhost`?

**Comment by michalwski** [comment]
If I understand it correctly, the `subhost` is f.e `muclight.esl.com` for the `host` `esl.com`.
There is function returning `host` by `subhost` (`gen_mod:get_module_and_host_by_subhost`). Is it possible that for `subhost` `muclight.esl.com`, this function returns sth else than `esl.com`?

File: apps/ejabberd/src/gen_mod.erl (hunk 1 above)
Comment (fenek) [inline apps/ejabberd/src/gen_mod.erl]: `subhost_mapping`? :)

File: apps/ejabberd/src/mod_mam_muc.erl (hunk 2 above)
Comment (fenek) [inline apps/ejabberd/src/mod_mam_muc.erl]: Sure, why not.

**Comment by fenek** [comment]
@michalwski 
Nope, hosts and subhosts have 1:n relation, so if the function returns something else than `esl.com` for `muclight.esl.com`, it's an error. Now when I think of it, module startup should fail if it tries to register under existing subdomain. I'll add it. Currently modules can poach subhost one from the other. :)

File: doc/modules/mod_mam.md [hunk 3]
```diff
@@ -22,7 +22,7 @@ If you haven't chosen any of the above, skip the next part.
 **Options:**
 
 * `mod_mam_muc`:
-    * `host` (optional, default: `"conference.@HOST@"`): MUC host that will be archived; **Warning:** If you are using MUC Light, make sure this option is set to MUC Light domain.
+    * `muc_module` (optional, default: `mod_muc`): MUC module the MAM service will attach to; **Warning:** If you are using MUC Light, make sure this option is set to `mod_muc_light`.
```
Comment (michalwski) [inline doc/modules/mod_mam.md]: What should we do if we use both `muc` and `muc_light`?

Comment (fenek) [inline doc/modules/mod_mam.md]: Not sure, is it even verified that `mod_mam_muc` can be started 2 times to support both `muc `and `muc_light`? If so, I think this change won't break anything because it just adds a translation step during init from module name to its subhost.

Comment (michalwski) [inline doc/modules/mod_mam.md]: This should be investigated then. A test cases where MAM works for both `mod_muc` and `mod_muc_light` would be great.

Comment (fenek) [inline doc/modules/mod_mam.md]: Could be a subject of another PR. Currently I won't have time probably to do any open source work. :(

**Comment by arcusfelis** [comment]
You can't start MAM modules twice (as any other modules).
But, yes, it's possible to extend mod_mam_muc to support several MUC modules at the same time.

**Comment by michalwski** [comment]
> You can't start MAM modules twice (as any other modules).

Actually I already started `mod_mam_muc` (the one before changes in this PR) twice on our staging server. Looks like everything works fine so far.

**Comment by fenek** [comment]
Will rebase, improve and open as new PR later.
